server/src/main/java/org/opensearch/index/mapper/WildcardFieldMapper.java

```diff
@@ -0,0 +1,825 @@
+            BooleanQuery.Builder builder = new BooleanQuery.Builder();
+            Set<String> expectedValues = new HashSet<>();
+            StringBuilder pattern = new StringBuilder();
+            for (Object value : values) {
+                String stringVal = value.toString();
+                builder.add(matchAllTermsQuery(name(), getRequiredNGrams(stringVal)), BooleanClause.Occur.SHOULD);
+                expectedValues.add(stringVal);
+                if (pattern.length() > 0) {
+                    pattern.append('|');
+                }
+                pattern.append(stringVal);
+            }
+            return new WildcardMatchingQuery(name(), builder.build(), expectedValues::contains, pattern.toString(), context, this);
+        }
+
+        private static BooleanQuery matchAllTermsQuery(String fieldName, Set<String> terms) {
+            BooleanQuery.Builder matchAllTermsBuilder = new BooleanQuery.Builder();
+            for (String term : terms) {
+                matchAllTermsBuilder.add(new TermQuery(new Term(fieldName, term)), BooleanClause.Occur.FILTER);
+            }
+            return matchAllTermsBuilder.build();
+        }
+    }
+
+    /**
+     * Custom two-phase query type for queries over the wildcard field. The expected behavior is that a first-phase
+     * query provides the best possible filter over the indexed trigrams, while the second phase matcher eliminates
+     * false positives by evaluating the true field value.
+     */
+    static class WildcardMatchingQuery extends Query {
+        private static final long MATCH_COST_ESTIMATE = 1000L;
+        private final String fieldName;
+        private final Query firstPhaseQuery;
+        private final Predicate<String> secondPhaseMatcher;
+        private final String patternString; // For toString
+        private final ValueFetcher valueFetcher;
+        private final SearchLookup searchLookup;
+
+        WildcardMatchingQuery(String fieldName, Query firstPhaseQuery, String patternString) {
+            this(fieldName, firstPhaseQuery, s -> true, patternString, (QueryShardContext) null, null);
+        }
+
+        public WildcardMatchingQuery(
+            String fieldName,
+            Query firstPhaseQuery,
+            Predicate<String> secondPhaseMatcher,
+            String patternString,
+            QueryShardContext context,
+            WildcardFieldType fieldType
+        ) {
+            this.fieldName = Objects.requireNonNull(fieldName);
+            this.firstPhaseQuery = Objects.requireNonNull(firstPhaseQuery);
+            this.secondPhaseMatcher = Objects.requireNonNull(secondPhaseMatcher);
+            this.patternString = Objects.requireNonNull(patternString);
+            if (context != null) {
+                this.searchLookup = context.lookup();
+                this.valueFetcher = fieldType.valueFetcher(context, context.lookup(), null);
+            } else {
+                this.searchLookup = null;
```

Comment:
Is this explicit setting to null required ? if we don't set anything, by default it will be null ?

Reply:
It's required since the fields are `final`, so all code paths must assign a value.

In practice, this branch is only for testing the query logic in the absence of a real index (so we don't have a `QueryShardContext`).